### ISSUE TITLE
In statement formatting

### ISSUE THREAD

**Issue by tomatualus** [Drizin/DapperQueryBuilder#45]
Hey. I am wondering if i am missing something, but i can't get array parameters to be formatted like so:

```csharp
private readonly List<string> _someValues = new List<string>()
{
    "value1",
    "value2",
    "value3"
};

var formattedValues = _someValues.Select(type => $"{type:varchar(50)}").ToList(); // <-- doesn't work because interpolated string expression is not IFormattable 

var query = connection.QueryBuilder("select something from somewhere";
query += @$"where table.values IN {formattedValues}";
```

The only solution i can think of is this:

```csharp
var formattedValues = _someValues.Select(type => new DbString
{
    Value = type,
    IsFixedLength = true,
    Length = 50,
    IsAnsi = true
})
```

**Comment by Drizin** [comment]
`var formattedValues = _someValues.Select(type => $"{type:varchar(50)}").ToList();`
Actually this doesn't work because the compiler by default will convert FormattableString to strings, so what you get is ` List<string>`, with plain values.

This works: `var formattedValues = _someValues.Select(type => (FormattableString) $"{type:varchar(50)}").ToList();`

Your other solution (using DbString) should be marginally faster, though, since it doesn't involve regex.

**Comment by tomatualus** [comment]
That throws this exception 

```
 at Dapper.SqlMapper.LookupDbType(Type type, String name, Boolean demand, ITypeHandler& handler) in /_/Dapper/SqlMapper.cs:line 427
[2022-07-26T06:48:27.716Z]    at Dapper.SqlMapper.LookupDbType(Type type, String name, Boolean demand, ITypeHandler& handler) in /_/Dapper/SqlMapper.cs:line 427
[2022-07-26T06:48:27.718Z]    at Dapper.SqlMapper.PackListParameters(IDbCommand command, String namePrefix, Object value) in /_/Dapper/SqlMapper.cs:line 2058
[2022-07-26T06:48:27.719Z]    at Dapper.SqlMapper.PackListParameters(IDbCommand command, String namePrefix, Object value) in /_/Dapper/SqlMapper.cs:line 2058
[2022-07-26T06:48:27.721Z]    at Dapper.DynamicParameters.AddParameters(IDbCommand command, Identity identity) in /_/Dapper/DynamicParameters.cs:line 248
[2022-07-26T06:48:27.722Z]    at Dapper.DynamicParameters.AddParameters(IDbCommand command, Identity identity) in /_/Dapper/DynamicParameters.cs:line 248
[2022-07-26T06:48:27.723Z]    at Dapper.DynamicParameters.Dapper.SqlMapper.IDynamicParameters.AddParameters(IDbCommand command, Identity identity) in /_/Dapper/DynamicParameters.cs:line 151
[2022-07-26T06:48:27.724Z]    at Dapper.DynamicParameters.Dapper.SqlMapper.IDynamicParameters.AddParameters(IDbCommand command, Identity identity) in /_/Dapper/DynamicParameters.cs:line 151
[2022-07-26T06:48:27.726Z]    at DapperQueryBuilder.ParameterInfos.Dapper.SqlMapper.IDynamicParameters.AddParameters(IDbCommand command, Identity identity)
[2022-07-26T06:48:27.727Z]    at DapperQueryBuilder.ParameterInfos.Dapper.SqlMapper.IDynamicParameters.AddParameters(IDbCommand command, Identity identity)
[2022-07-26T06:48:27.728Z]    at Dapper.SqlMapper.<>c__DisplayClass165_0.<GetCacheInfo>b__0(IDbCommand cmd, Object obj) in /_/Dapper/SqlMapper.cs:line 1738
[2022-07-26T06:48:27.729Z]    at Dapper.SqlMapper.<>c__DisplayClass165_0.<GetCacheInfo>b__0(IDbCommand cmd, Object obj) in /_/Dapper/SqlMapper.cs:line 1738
[2022-07-26T06:48:27.730Z]    at Dapper.CommandDefinition.SetupCommand(IDbConnection cnn, Action`2 paramReader) in /_/Dapper/CommandDefinition.cs:line 129
[2022-07-26T06:48:27.731Z]    at Dapper.CommandDefinition.SetupCommand(IDbConnection cnn, Action`2 paramReader) in /_/Dapper/CommandDefinition.cs:line 129
[2022-07-26T06:48:27.734Z]    at Dapper.SqlMapper.TrySetupAsyncCommand(CommandDefinition command, IDbConnection cnn, Action`2 paramReader) in /_/Dapper/SqlMapper.Async.cs:line 396
[2022-07-26T06:48:27.736Z]    at Dapper.SqlMapper.TrySetupAsyncCommand(CommandDefinition command, IDbConnection cnn, Action`2 paramReader) in /_/Dapper/SqlMapper.Async.cs:line 396
[2022-07-26T06:48:27.737Z]    at Dapper.SqlMapper.QueryAsync[T](IDbConnection cnn, Type effectiveType, CommandDefinition command) in /_/Dapper/SqlMapper.Async.cs:line 413
[2022-07-26T06:48:27.739Z]    at Dapper.SqlMapper.QueryAsync[T](IDbConnection cnn, Type effectiveType, CommandDefinition command) in /_/Dapper/SqlMapper.Async.cs:line 413

```

**Comment by Drizin** [comment]
I suppose it's because [InterpolatedStatementParser](https://github.com/Drizin/DapperQueryBuilder/blob/master/src/DapperQueryBuilder/InterpolatedStatementParser.cs#L224) just forwards any IEnumerable to Dapper, and Dapper doesn't understand that format (a FormattableString type with a format specifying the dbtype).
To do what you expect we (InterpolatedStatementParser.cs) would have to unwrap the `IEnumerable<FormattableString>`, and convert it into `IEnumerable<DbString>` which would then be passed to Dapper (like you did - that worked, right?).

**Comment by tomatualus** [comment]
Yes, converting it into a `IEnumerable<DbString>` worked, but that takes too much lines of code :(

**Comment by Drizin** [comment]
I guess it doesn't make sense to provide a `string ":dbtype"` for **each** element of the `IEnumerable`, but instead a more intuitive approach would be to define the dbtype for all array elements at once like this:
```cs
List<string> _someValues = new List<string>() { "value1", "value2", "value" };
var query = cn.QueryBuilder($"select something from somewhere");
query += $@"where table.values IN {_someValues:varchar(50)}";
```
In InterpolatedStatementParser we would check if arg is `IEnumerable<string>` (only for the [fixed length](https://github.com/Drizin/DapperQueryBuilder/blob/master/src/DapperQueryBuilder/InterpolatedStatementParser.cs#L182-L196) strings, and convert arg from `IEnumerable<string>` to `IEnumerable<DbString>`.

Does that sound right?

**Comment by Drizin** [comment]
Please check my [last commit](https://github.com/Drizin/DapperQueryBuilder/commit/ef7cfb4f88a62ae27ad3cf59a2a2a5537ea3a284) - let me know if it works fine.
And thanks for the nice suggestion!

**Comment by tomatualus** [comment]
Hey. Looks good to me :)

**Comment by Drizin** [comment]
Sounds good! Published in [Dapper-QueryBuilder Nuget 1.2.6](https://www.nuget.org/packages/Dapper-QueryBuilder/1.2.6)
Thanks again!
